doc/run-cloud-build-locally.md

```diff
@@ -161,37 +158,221 @@ Execute the HAPI search test:
 ```
 
-Repeat for each mode (JDBC, Bulk Export, controller/spark, OpenMRS) using the
-matching environment variables from the original `cloudbuild.yaml`.
+## JDBC Mode
+
+**Note:** Only run this section if debugging JDBC mode; otherwise skip.
+
+### 1. Run batch pipeline for JDBC mode
+
+Run the JDBC mode job against HAPI:
+
+```bash
+docker run --rm \
+  -e JDBC_MODE_ENABLED=true \
+  -e JDBC_MODE_HAPI=true \
+  -e FHIR_SERVER_URL=http://localhost:9001/fhir \
+  -e SINK_PATH=http://localhost:9002/fhir \
+  -e SINK_USERNAME=hapi -e SINK_PASSWORD=hapi \
+  -e FHIR_DATABASE_CONFIG_PATH=/workspace/utils/hapi-postgres-config.json \
+  -e PARQUET_PATH=/workspace/e2e-tests/JDBC_HAPI \
+  -e OUTPUT_PARQUET_VIEW_PATH=/workspace/e2e-tests/JDBC_HAPI/VIEWS_TIMESTAMP_1 \
+  -e JDBC_FETCH_SIZE=1000 \
+  -v $(pwd):/workspace \
+  --network host \
+  ${_REPOSITORY}/batch-pipeline:${_TAG}
+```
+
+### 2. Run e2e tests for JDBC mode
+
+```bash
+./e2e-tests/pipeline_validation.sh e2e-tests/ JDBC_HAPI JDBC_HAPI_FHIR_JSON http://localhost:9002/fhir
+```
+
+## Bulk Export Mode
+
+**Note:** Only run this section if debugging Bulk Export mode; otherwise skip.
+
+### 1. Run batch pipeline for Bulk Export mode
+
+```bash
+docker run --rm \
+  -e FHIR_FETCH_MODE=BULK_EXPORT \
+  -e FHIR_SERVER_URL=http://localhost:9001/fhir \
+  -e PARQUET_PATH=/workspace/e2e-tests/BULK_EXPORT \
+  -e OUTPUT_PARQUET_VIEW_PATH=/workspace/e2e-tests/BULK_EXPORT/VIEWS_TIMESTAMP_1 \
+  -v $(pwd):/workspace \
+  --network host \
+  ${_REPOSITORY}/batch-pipeline:${_TAG}
+```
+
+### 2. Run e2e tests for Bulk Export mode
+
+```bash
+./e2e-tests/pipeline_validation.sh e2e-tests/ BULK_EXPORT BULK_EXPORT_FHIR_JSON
+```
+
+## Controller and Spark Mode
 
-### 7. Clean up containers
+**Note:** Only run this section if debugging Controller and Spark mode;
+otherwise skip.
 
-When finished, bring everything down to avoid port conflicts:
+### 1. Turn down sink servers
 
 ```bash
 docker compose -f docker/sink-compose.yml -p sink-server-search down -v
 docker compose -f docker/sink-compose.yml -p sink-server-jdbc down -v
-docker compose -f docker/hapi-compose.yml -p hapi-compose down -v
 ```
 
-Also stop the controller/spark Compose services and other environment-specific
-stacks as needed.
+### 2. Create views database
+
+The `views` database is used for creating flat views from ViewDefinitions.
+
+```bash
+docker run --rm --network host postgres psql -U admin -d postgres -h localhost -p 5432 -c 'CREATE DATABASE views;'
+```
+
+### 3. Build the e2e controller/spark image
+
+```bash
+docker build -t ${_REPOSITORY}/e2e-tests/controller-spark:${_TAG} -f e2e-tests/controller-spark/Dockerfile .
+```
+
+### 4. Bring up controller and Spark containers
+
+```bash
+PIPELINE_CONFIG=/workspace/docker/config DWH_ROOT=/workspace/e2e-tests/controller-spark/dwh docker compose -f docker/compose-controller-spark-sql-single.yaml up --force-recreate -d
```

Comment:
Don't you need a volume mapping for `/workspace`?
nit: Also please  break the long line (didn't our default formatter catch this?).

Reply:
The bind mount is defined in the compose config file here - https://github.com/google/fhir-data-pipes/blob/547612dea431ebec8800fb99f9674150e06dce57/docker/compose-controller-spark-sql-single.yaml#L60-L62

Reply:
Don't think our default formatter(Spotless) can reformat these commands as appears in the `.md` files. We probably need to break these down manually.